File: daedalos-tools/codex/src/codex/chunker.py
"""Code chunking for semantic search."""

import re
from dataclasses import dataclass
from typing import List, Optional
from pathlib import Path


@dataclass
class CodeChunk:
    """A chunk of code for indexing."""
    file_path: str
    start_line: int
    end_line: int
    content: str
    chunk_type: str  # function, class, block, file
    name: str  # function/class name if applicable

    @property
    def location(self) -> str:
        return f"{self.file_path}:{self.start_line}"

    def __str__(self) -> str:
        return f"[{self.chunk_type}] {self.name} at {self.location}"


class CodeChunker:
    """Chunk code files into searchable segments."""

    # Extensions and their chunkers
    CHUNKERS = {
        "py": "_chunk_python",
        "js": "_chunk_javascript",
        "ts": "_chunk_javascript",
        "tsx": "_chunk_javascript",
        "jsx": "_chunk_javascript",
        "swift": "_chunk_swift",
        "rs": "_chunk_rust",
        "go": "_chunk_go",
        "rb": "_chunk_ruby",
    }

    def chunk_file(self, path: str, content: str) -> List[CodeChunk]:
        """Chunk a file into searchable segments."""
        ext = path.split(".")[-1].lower()
        chunker_name = self.CHUNKERS.get(ext)

        if chunker_name:
            chunker = getattr(self, chunker_name)
            chunks = chunker(path, content)
        else:
            chunks = self._chunk_generic(path, content)

        # Always include whole file as a chunk for context
        if content.strip():
            chunks.append(CodeChunk(
                file_path=path,
                start_line=1,
                end_line=content.count("\n") + 1,
                content=content[:2000],  # Truncate for large files
                chunk_type="file",
                name=Path(path).name
            ))

        return chunks

    def _chunk_python(self, path: str, content: str) -> List[CodeChunk]:
        """Chunk Python code."""
        chunks = []
        lines = content.split("\n")

        # Pattern for function and class definitions
        pattern = r"^(class|def|async def)\s+(\w+)"
        current_chunk = None
        current_indent = 0

        for i, line in enumerate(lines, 1):
            match = re.match(pattern, line)
            if match:
                # Save previous chunk
                if current_chunk:
                    current_chunk.end_line = i - 1
                    current_chunk.content = "\n".join(
                        lines[current_chunk.start_line - 1:current_chunk.end_line]
                    )
                    chunks.append(current_chunk)

                chunk_type = "class" if match.group(1) == "class" else "function"
                current_chunk = CodeChunk(
                    file_path=path,
                    start_line=i,
                    end_line=i,
                    content="",
                    chunk_type=chunk_type,
                    name=match.group(2)
                )
                current_indent = len(line) - len(line.lstrip())

        # Don't forget the last chunk
        if current_chunk:
            current_chunk.end_line = len(lines)
            current_chunk.content = "\n".join(
                lines[current_chunk.start_line - 1:current_chunk.end_line]
            )
            chunks.append(current_chunk)

        return chunks

    def _chunk_javascript(self, path: str, content: str) -> List[CodeChunk]:
        """Chunk JavaScript/TypeScript code."""
        chunks = []
        lines = content.split("\n")

        # Patterns for functions, classes, and exports
        patterns = [
            (r"^(export\s+)?(async\s+)?function\s+(\w+)", "function", 3),
            (r"^(export\s+)?class\s+(\w+)", "class", 2),
            (r"^(export\s+)?const\s+(\w+)\s*=\s*(async\s+)?\(", "function", 2),
            (r"^(export\s+)?const\s+(\w+)\s*=\s*(async\s+)?function", "function", 2),
        ]

        current_chunk = None
        brace_depth = 0

        for i, line in enumerate(lines, 1):
            for pattern, chunk_type, name_group in patterns:
                match = re.match(pattern, line.strip())
                if match:
                    if current_chunk:
                        current_chunk.end_line = i - 1
                        current_chunk.content = "\n".join(
                            lines[current_chunk.start_line - 1:current_chunk.end_line]
                        )
                        chunks.append(current_chunk)

                    current_chunk = CodeChunk(
                        file_path=path,
                        start_line=i,
                        end_line=i,
                        content="",
                        chunk_type=chunk_type,
                        name=match.group(name_group)
                    )
                    break

        if current_chunk:
            current_chunk.end_line = len(lines)
            current_chunk.content = "\n".join(
                lines[current_chunk.start_line - 1:current_chunk.end_line]
            )
            chunks.append(current_chunk)

        return chunks

    def _chunk_swift(self, path: str, content: str) -> List[CodeChunk]:
        """Chunk Swift code."""
        chunks = []
        lines = content.split("\n")

        patterns = [
            (r"^\s*(public|private|internal|open)?\s*func\s+(\w+)", "function", 2),
            (r"^\s*(public|private|internal|open)?\s*class\s+(\w+)", "class", 2),
            (r"^\s*(public|private|internal|open)?\s*struct\s+(\w+)", "struct", 2),
            (r"^\s*(public|private|internal|open)?\s*enum\s+(\w+)", "enum", 2),
            (r"^\s*(public|private|internal|open)?\s*protocol\s+(\w+)", "protocol", 2),
        ]

        current_chunk = None

        for i, line in enumerate(lines, 1):
            for pattern, chunk_type, name_group in patterns:
                match = re.match(pattern, line)
                if match:
                    if current_chunk:
                        current_chunk.end_line = i - 1
                        current_chunk.content = "\n".join(
                            lines[current_chunk.start_line - 1:current_chunk.end_line]
                        )
                        chunks.append(current_chunk)

                    current_chunk = CodeChunk(
                        file_path=path,
                        start_line=i,
                        end_line=i,
                        content="",
                        chunk_type=chunk_type,
                        name=match.group(name_group)
                    )
                    break

        if current_chunk:
            current_chunk.end_line = len(lines)
            current_chunk.content = "\n".join(
                lines[current_chunk.start_line - 1:current_chunk.end_line]
            )
            chunks.append(current_chunk)

        return chunks

    def _chunk_rust(self, path: str, content: str) -> List[CodeChunk]:
        """Chunk Rust code."""
        chunks = []
        lines = content.split("\n")

        patterns = [
            (r"^\s*(pub\s+)?fn\s+(\w+)", "function", 2),
            (r"^\s*(pub\s+)?struct\s+(\w+)", "struct", 2),
            (r"^\s*(pub\s+)?enum\s+(\w+)", "enum", 2),
            (r"^\s*(pub\s+)?trait\s+(\w+)", "trait", 2),
            (r"^\s*impl\s+(\w+)", "impl", 1),
        ]

        current_chunk = None

        for i, line in enumerate(lines, 1):
            for pattern, chunk_type, name_group in patterns:
                match = re.match(pattern, line)
                if match:
                    if current_chunk:
                        current_chunk.end_line = i - 1
                        current_chunk.content = "\n".join(
                            lines[current_chunk.start_line - 1:current_chunk.end_line]
                        )
                        chunks.append(current_chunk)

                    current_chunk = CodeChunk(
                        file_path=path,
                        start_line=i,
                        end_line=i,
                        content="",
                        chunk_type=chunk_type,
                        name=match.group(name_group)
                    )
                    break

        if current_chunk:
            current_chunk.end_line = len(lines)
            current_chunk.content = "\n".join(
                lines[current_chunk.start_line - 1:current_chunk.end_line]
            )
            chunks.append(current_chunk)

        return chunks

    def _chunk_go(self, path: str, content: str) -> List[CodeChunk]:
        """Chunk Go code."""
        chunks = []
        lines = content.split("\n")

        patterns = [
            (r"^func\s+(\w+)", "function", 1),
            (r"^func\s+\([^)]+\)\s+(\w+)", "method", 1),
            (r"^type\s+(\w+)\s+struct", "struct", 1),
            (r"^type\s+(\w+)\s+interface", "interface", 1),
        ]

        current_chunk = None

        for i, line in enumerate(lines, 1):
            for pattern, chunk_type, name_group in patterns:
                match = re.match(pattern, line)
                if match:
                    if current_chunk:
                        current_chunk.end_line = i - 1
                        current_chunk.content = "\n".join(
                            lines[current_chunk.start_line - 1:current_chunk.end_line]
                        )
                        chunks.append(current_chunk)

                    current_chunk = CodeChunk(
                        file_path=path,
                        start_line=i,
                        end_line=i,
                        content="",
                        chunk_type=chunk_type,
                        name=match.group(name_group)
                    )
                    break

        if current_chunk:
            current_chunk.end_line = len(lines)
            current_chunk.content = "\n".join(
                lines[current_chunk.start_line - 1:current_chunk.end_line]
            )
            chunks.append(current_chunk)

        return chunks

    def _chunk_ruby(self, path: str, content: str) -> List[CodeChunk]:
        """Chunk Ruby code."""
        chunks = []
        lines = content.split("\n")

        patterns = [
            (r"^\s*def\s+(\w+)", "method", 1),
            (r"^\s*class\s+(\w+)", "class", 1),
            (r"^\s*module\s+(\w+)", "module", 1),
        ]

        current_chunk = None

        for i, line in enumerate(lines, 1):
            for pattern, chunk_type, name_group in patterns:
                match = re.match(pattern, line)
                if match:
                    if current_chunk:
                        current_chunk.end_line = i - 1
                        current_chunk.content = "\n".join(
                            lines[current_chunk.start_line - 1:current_chunk.end_line]
                        )
                        chunks.append(current_chunk)

                    current_chunk = CodeChunk(
                        file_path=path,
                        start_line=i,
                        end_line=i,
                        content="",
                        chunk_type=chunk_type,
                        name=match.group(name_group)
                    )
                    break

        if current_chunk:
            current_chunk.end_line = len(lines)
            current_chunk.content = "\n".join(
                lines[current_chunk.start_line - 1:current_chunk.end_line]
            )
            chunks.append(current_chunk)

        return chunks

    def _chunk_generic(self, path: str, content: str) -> List[CodeChunk]:
        """Generic chunking for unknown file types."""
        # Just split into blocks of ~50 lines
        chunks = []
        lines = content.split("\n")
        chunk_size = 50

        for i in range(0, len(lines), chunk_size):
            chunk_lines = lines[i:i + chunk_size]
            chunks.append(CodeChunk(
                file_path=path,
                start_line=i + 1,
                end_line=min(i + chunk_size, len(lines)),
                content="\n".join(chunk_lines),
                chunk_type="block",
                name=f"block_{i // chunk_size + 1}"
            ))

        return chunks

File: crates/codex/src/chunker.rs
//! Code chunking for semantic search.
//!
//! Chunks code files into searchable segments based on semantic units
//! (functions, classes, structs) rather than fixed-size blocks.

use regex::Regex;
use std::path::Path;

/// A chunk of code for indexing
#[derive(Debug, Clone)]
pub struct CodeChunk {
    /// Path relative to project root
    pub file_path: String,
    /// Starting line number (1-indexed)
    pub start_line: usize,
    /// Ending line number (1-indexed)
    pub end_line: usize,
    /// The code content
    pub content: String,
    /// Type of chunk (function, class, struct, file, block)
    pub chunk_type: String,
    /// Name of the function/class/etc
    pub name: String,
}

impl CodeChunk {
    /// Get the file:line location string
    pub fn location(&self) -> String {
        format!("{}:{}", self.file_path, self.start_line)
    }
}

/// Chunker that splits code into semantic units
pub struct CodeChunker {
    // Compiled patterns for different languages
    python_pattern: Regex,
    rust_fn_pattern: Regex,
    rust_struct_pattern: Regex,
    rust_enum_pattern: Regex,
    rust_trait_pattern: Regex,
    rust_impl_pattern: Regex,
    js_fn_pattern: Regex,
    js_class_pattern: Regex,
    js_const_fn_pattern: Regex,
    go_fn_pattern: Regex,
    go_method_pattern: Regex,
    go_struct_pattern: Regex,
    go_interface_pattern: Regex,
}

impl Default for CodeChunker {
    fn default() -> Self {
        Self::new()
    }
}

impl CodeChunker {
    pub fn new() -> Self {
        Self {
            python_pattern: Regex::new(r"^(class|def|async def)\s+(\w+)").unwrap(),
            rust_fn_pattern: Regex::new(r"^\s*(pub\s+)?fn\s+(\w+)").unwrap(),
            rust_struct_pattern: Regex::new(r"^\s*(pub\s+)?struct\s+(\w+)").unwrap(),
            rust_enum_pattern: Regex::new(r"^\s*(pub\s+)?enum\s+(\w+)").unwrap(),
            rust_trait_pattern: Regex::new(r"^\s*(pub\s+)?trait\s+(\w+)").unwrap(),
            rust_impl_pattern: Regex::new(r"^\s*impl\s+(?:<[^>]+>\s+)?(\w+)").unwrap(),
            js_fn_pattern: Regex::new(r"^(export\s+)?(async\s+)?function\s+(\w+)").unwrap(),
            js_class_pattern: Regex::new(r"^(export\s+)?class\s+(\w+)").unwrap(),
            js_const_fn_pattern: Regex::new(r"^(export\s+)?const\s+(\w+)\s*=\s*(async\s+)?(\(|function)").unwrap(),
            go_fn_pattern: Regex::new(r"^func\s+(\w+)").unwrap(),
            go_method_pattern: Regex::new(r"^func\s+\([^)]+\)\s+(\w+)").unwrap(),
            go_struct_pattern: Regex::new(r"^type\s+(\w+)\s+struct").unwrap(),
            go_interface_pattern: Regex::new(r"^type\s+(\w+)\s+interface").unwrap(),
        }
    }

    /// Chunk a file into searchable segments
    pub fn chunk_file(&self, path: &str, content: &str) -> Vec<CodeChunk> {
        let ext = Path::new(path)
            .extension()
            .and_then(|e| e.to_str())
            .unwrap_or("")
            .to_lowercase();

        let mut chunks = match ext.as_str() {
            "py" | "pyw" => self.chunk_python(path, content),
            "rs" => self.chunk_rust(path, content),
            "js" | "jsx" | "mjs" | "cjs" | "ts" | "tsx" | "mts" | "cts" => {
                self.chunk_javascript(path, content)
            }
            "go" => self.chunk_go(path, content),
            _ => self.chunk_generic(path, content),
        };

        // Always include whole file as a chunk for context
        if !content.trim().is_empty() {
            let line_count = content.lines().count();
            let truncated = if content.len() > 2000 {
                &content[..2000]
            } else {
                content
            };
            chunks.push(CodeChunk {
                file_path: path.to_string(),
                start_line: 1,
                end_line: line_count.max(1),
                content: truncated.to_string(),
                chunk_type: "file".to_string(),
                name: Path::new(path)
                    .file_name()
                    .and_then(|n| n.to_str())
                    .unwrap_or(path)
                    .to_string(),
            });
        }

        chunks
    }

    fn chunk_python(&self, path: &str, content: &str) -> Vec<CodeChunk> {
        let lines: Vec<&str> = content.lines().collect();
        let mut chunks = Vec::new();
        let mut current_chunk: Option<(usize, String, String)> = None; // (start_line, type, name)

        for (i, line) in lines.iter().enumerate() {
            let line_num = i + 1;
            if let Some(captures) = self.python_pattern.captures(line) {
                // Save previous chunk
                if let Some((start, chunk_type, name)) = current_chunk.take() {
                    let chunk_content = lines[start - 1..i].join("\n");
                    chunks.push(CodeChunk {
                        file_path: path.to_string(),
                        start_line: start,
                        end_line: i,
                        content: chunk_content,
                        chunk_type,
                        name,
                    });
                }

                let keyword = captures.get(1).map(|m| m.as_str()).unwrap_or("");
                let name = captures.get(2).map(|m| m.as_str()).unwrap_or("unknown");
                let chunk_type = if keyword == "class" {
                    "class"
                } else {
                    "function"
                };

                current_chunk = Some((line_num, chunk_type.to_string(), name.to_string()));
            }
        }

        // Don't forget the last chunk
        if let Some((start, chunk_type, name)) = current_chunk {
            let chunk_content = lines[start - 1..].join("\n");
            chunks.push(CodeChunk {
                file_path: path.to_string(),
                start_line: start,
                end_line: lines.len(),
                content: chunk_content,
                chunk_type,
                name,
            });
        }

        chunks
    }

    fn chunk_rust(&self, path: &str, content: &str) -> Vec<CodeChunk> {
        let lines: Vec<&str> = content.lines().collect();
        let mut chunks = Vec::new();
        let mut current_chunk: Option<(usize, String, String)> = None;

        for (i, line) in lines.iter().enumerate() {
            let line_num = i + 1;

            // Try each pattern
            let match_result = if let Some(captures) = self.rust_fn_pattern.captures(line) {
                Some(("function", captures.get(2).map(|m| m.as_str()).unwrap_or("unknown")))
            } else if let Some(captures) = self.rust_struct_pattern.captures(line) {
                Some(("struct", captures.get(2).map(|m| m.as_str()).unwrap_or("unknown")))
            } else if let Some(captures) = self.rust_enum_pattern.captures(line) {
                Some(("enum", captures.get(2).map(|m| m.as_str()).unwrap_or("unknown")))
            } else if let Some(captures) = self.rust_trait_pattern.captures(line) {
                Some(("trait", captures.get(2).map(|m| m.as_str()).unwrap_or("unknown")))
            } else if let Some(captures) = self.rust_impl_pattern.captures(line) {
                Some(("impl", captures.get(1).map(|m| m.as_str()).unwrap_or("unknown")))
            } else {
                None
            };

            if let Some((chunk_type, name)) = match_result {
                // Save previous chunk
                if let Some((start, prev_type, prev_name)) = current_chunk.take() {
                    let chunk_content = lines[start - 1..i].join("\n");
                    chunks.push(CodeChunk {
                        file_path: path.to_string(),
                        start_line: start,
                        end_line: i,
                        content: chunk_content,
                        chunk_type: prev_type,
                        name: prev_name,
                    });
                }

                current_chunk = Some((line_num, chunk_type.to_string(), name.to_string()));
            }
        }

        // Don't forget the last chunk
        if let Some((start, chunk_type, name)) = current_chunk {
            let chunk_content = lines[start - 1..].join("\n");
            chunks.push(CodeChunk {
                file_path: path.to_string(),
                start_line: start,
                end_line: lines.len(),
                content: chunk_content,
                chunk_type,
                name,
            });
        }

        chunks
    }

    fn chunk_javascript(&self, path: &str, content: &str) -> Vec<CodeChunk> {
        let lines: Vec<&str> = content.lines().collect();
        let mut chunks = Vec::new();
        let mut current_chunk: Option<(usize, String, String)> = None;

        for (i, line) in lines.iter().enumerate() {
            let line_num = i + 1;
            let trimmed = line.trim();

            // Try each pattern
            let match_result = if let Some(captures) = self.js_fn_pattern.captures(trimmed) {
                Some(("function", captures.get(3).map(|m| m.as_str()).unwrap_or("unknown")))
            } else if let Some(captures) = self.js_class_pattern.captures(trimmed) {
                Some(("class", captures.get(2).map(|m| m.as_str()).unwrap_or("unknown")))
            } else if let Some(captures) = self.js_const_fn_pattern.captures(trimmed) {
                Some(("function", captures.get(2).map(|m| m.as_str()).unwrap_or("unknown")))
            } else {
                None
            };

            if let Some((chunk_type, name)) = match_result {
                // Save previous chunk
                if let Some((start, prev_type, prev_name)) = current_chunk.take() {
                    let chunk_content = lines[start - 1..i].join("\n");
                    chunks.push(CodeChunk {
                        file_path: path.to_string(),
                        start_line: start,
                        end_line: i,
                        content: chunk_content,
                        chunk_type: prev_type,
                        name: prev_name,
                    });
                }

                current_chunk = Some((line_num, chunk_type.to_string(), name.to_string()));
            }
        }

        // Don't forget the last chunk
        if let Some((start, chunk_type, name)) = current_chunk {
            let chunk_content = lines[start - 1..].join("\n");
            chunks.push(CodeChunk {
                file_path: path.to_string(),
                start_line: start,
                end_line: lines.len(),
                content: chunk_content,
                chunk_type,
                name,
            });
        }

        chunks
    }

    fn chunk_go(&self, path: &str, content: &str) -> Vec<CodeChunk> {
        let lines: Vec<&str> = content.lines().collect();
        let mut chunks = Vec::new();
        let mut current_chunk: Option<(usize, String, String)> = None;

        for (i, line) in lines.iter().enumerate() {
            let line_num = i + 1;

            // Try each pattern
            let match_result = if let Some(captures) = self.go_fn_pattern.captures(line) {
                Some(("function", captures.get(1).map(|m| m.as_str()).unwrap_or("unknown")))
            } else if let Some(captures) = self.go_method_pattern.captures(line) {
                Some(("method", captures.get(1).map(|m| m.as_str()).unwrap_or("unknown")))
            } else if let Some(captures) = self.go_struct_pattern.captures(line) {
                Some(("struct", captures.get(1).map(|m| m.as_str()).unwrap_or("unknown")))
            } else if let Some(captures) = self.go_interface_pattern.captures(line) {
                Some(("interface", captures.get(1).map(|m| m.as_str()).unwrap_or("unknown")))
            } else {
                None
            };

            if let Some((chunk_type, name)) = match_result {
                // Save previous chunk
                if let Some((start, prev_type, prev_name)) = current_chunk.take() {
                    let chunk_content = lines[start - 1..i].join("\n");
                    chunks.push(CodeChunk {
                        file_path: path.to_string(),
                        start_line: start,
                        end_line: i,
                        content: chunk_content,
                        chunk_type: prev_type,
                        name: prev_name,
                    });
                }

                current_chunk = Some((line_num, chunk_type.to_string(), name.to_string()));
            }
        }

        // Don't forget the last chunk
        if let Some((start, chunk_type, name)) = current_chunk {
            let chunk_content = lines[start - 1..].join("\n");
            chunks.push(CodeChunk {
                file_path: path.to_string(),
                start_line: start,
                end_line: lines.len(),
                content: chunk_content,
                chunk_type,
                name,
            });
        }

        chunks
    }

    fn chunk_generic(&self, path: &str, content: &str) -> Vec<CodeChunk> {
        // Split into blocks of ~50 lines for unknown file types
        let lines: Vec<&str> = content.lines().collect();
        let chunk_size = 50;
        let mut chunks = Vec::new();

        for (block_num, chunk_lines) in lines.chunks(chunk_size).enumerate() {
            let start = block_num * chunk_size + 1;
            let end = start + chunk_lines.len() - 1;

            chunks.push(CodeChunk {
                file_path: path.to_string(),
                start_line: start,
                end_line: end,
                content: chunk_lines.join("\n"),
                chunk_type: "block".to_string(),
                name: format!("block_{}", block_num + 1),
            });
        }

        chunks
    }
}
